File: src/types/dendro-types.ts
import validateObject, { isArrayOf, isBoolean, isEqualTo, isNumber, isOneOf, isString, optional, isNull } from "./validateObject"

export type DendroProjectUser = {
    userId: string
    role: 'admin' | 'editor' | 'viewer'
}

export const isDendroProjectUser = (x: any): x is DendroProjectUser => {
    return validateObject(x, {
        userId: isString,
        role: isOneOf([isEqualTo('admin'), isEqualTo('editor'), isEqualTo('viewer')])
    })
}

export type DendroProject = {
    projectId: string
    name: string
    description: string
    ownerId: string
    users: DendroProjectUser[]
    publiclyReadable: boolean
    computeResourceId?: string | null
    tags: string[]
    timestampCreated: number
    timestampModified: number
}

export const isDendroProject = (x: any): x is DendroProject => {
    return validateObject(x, {
        projectId: isString,
        name: isString,
        description: isString,
        ownerId: isString,
        users: isArrayOf(isDendroProjectUser),
        publiclyReadable: isBoolean,
        computeResourceId: optional(isOneOf([isString, isNull])),
        tags: isArrayOf(isString),
        timestampCreated: isNumber,
        timestampModified: isNumber
    })
}

export type DendroJobInputFile = {
    name: string
    fileId: string
    fileName: string
}

export const isDendroJobInputFile = (x: any): x is DendroJobInputFile => {
    return validateObject(x, {
        name: isString,
        fileId: isString,
        fileName: isString
    })
}

export type DendroJobInputParameter = {
    name: string
    value?: any
    secret?: boolean
}

export const isDendroJobInputParameter = (x: any): x is DendroJobInputParameter => {
    return validateObject(x, {
        name: isString,
        value: optional(isString),
        secret: optional(isBoolean)
    })
}

export type DendroJobOutputFile = {
    name: string
    fileName: string
    fileId?: string
}

export const isDendroJobOutputFile = (x: any): x is DendroJobOutputFile => {
    return validateObject(x, {
        name: isString,
        fileName: isString,
        fileId: optional(isString)
    })
}

export type ComputeResourceSpecProcessorParameter = {
    name: string
    help: string
    type: string
    default?: any
    options?: string[] | number[]
    secret: boolean
}

export const isComputeResourceSpecProcessorParameter = (x: any): x is ComputeResourceSpecProcessorParameter => {
    return validateObject(x, {
        name: isString,
        help: isString,
        type: isString,
        default: optional(isString),
        options: optional(isArrayOf(isOneOf([isString, isNumber]))),
        secret: isBoolean
    })
}

export type ComputeResourceSpecProcessorInput = {
    name: string
    help: string
    list: boolean
}

export const isComputeResourceSpecProcessorInput = (x: any): x is ComputeResourceSpecProcessorInput => {
    return validateObject(x, {
        name: isString,
        help: isString,
        list: isBoolean
    })
}

export type ComputeResourceSpecProcessorOutput = {
    name: string
    help: string
}

export const isComputeResourceSpecProcessorOutput = (x: any): x is ComputeResourceSpecProcessorOutput => {
    return validateObject(x, {
        name: isString,
        help: isString
    })
}

export type ComputeResourceSpecProcessorAttribute = {
    name: string
    value: any
}

export const isComputeResourceSpecProcessorAttribute = (x: any): x is ComputeResourceSpecProcessorAttribute => {
    return validateObject(x, {
        name: isString,
        value: isString
    })
}

export type ComputeResourceSpecProcessorTag = {
    tag: string
}

export const isComputeResourceSpecProcessorTag = (x: any): x is ComputeResourceSpecProcessorTag => {
    return validateObject(x, {
        tag: isString
    })
}

export type ComputeResourceSpecProcessor = {
    name: string
    help: string
    inputs: ComputeResourceSpecProcessorInput[]
    outputs: ComputeResourceSpecProcessorOutput[]
    parameters: ComputeResourceSpecProcessorParameter[]
    attributes: ComputeResourceSpecProcessorAttribute[]
    tags: ComputeResourceSpecProcessorTag[]
}

export const isComputeResourceSpecProcessor = (x: any): x is ComputeResourceSpecProcessor => {
    return validateObject(x, {
        name: isString,
        help: isString,
        inputs: isArrayOf(isComputeResourceSpecProcessorInput),
        outputs: isArrayOf(isComputeResourceSpecProcessorOutput),
        parameters: isArrayOf(isComputeResourceSpecProcessorParameter),
        attributes: isArrayOf(isComputeResourceSpecProcessorAttribute),
        tags: isArrayOf(isComputeResourceSpecProcessorTag)
    })
}

export type DendroJob = {
    projectId: string
    jobId: string
    jobPrivateKey: string
    userId: string
    processorName: string
    batchId?: string
    inputFiles: DendroJobInputFile[]
    inputFileIds: string[]
    inputParameters: DendroJobInputParameter[]
    outputFiles: DendroJobOutputFile[]
    timestampCreated: number
    computeResourceId: string
    status: 'pending' | 'queued' | 'starting' | 'running' | 'completed' | 'failed'
    error?: string
    processorVersion?: string
    computeResourceNodeId?: string
    computeResourceNodeName?: string
    consoleOutputUrl?: string
    timestampQueued?: number
    timestampStarting?: number
    timestampStarted?: number
    timestampFinished?: number
    outputFileIds?: string[]
    processorSpec: ComputeResourceSpecProcessor
    dandiApiKey?: string
}

export const isDendroJob = (x: any): x is DendroJob => {
    return validateObject(x, {
        projectId: isString,
        jobId: isString,
        jobPrivateKey: isString,
        userId: isString,
        processorName: isString,
        batchId: optional(isString),
        inputFiles: isArrayOf(isDendroJobInputFile),
        inputFileIds: isArrayOf(isString),
        inputParameters: isArrayOf(isDendroJobInputParameter),
        outputFiles: isArrayOf(isDendroJobOutputFile),
        timestampCreated: isNumber,
        computeResourceId: isString,
        status: isOneOf([isEqualTo('pending'), isEqualTo('queued'), isEqualTo('starting'), isEqualTo('running'), isEqualTo('completed'), isEqualTo('failed')]),
        error: optional(isString),
        processorVersion: optional(isString),
        computeResourceNodeId: optional(isString),
        computeResourceNodeName: optional(isString),
        consoleOutputUrl: optional(isString),
        timestampQueued: optional(isNumber),
        timestampStarting: optional(isNumber),
        timestampStarted: optional(isNumber),
        timestampFinished: optional(isNumber),
        outputFileIds: optional(isArrayOf(isString)),
        processorSpec: isComputeResourceSpecProcessor,
        dandiApiKey: optional(isString)
    })
}

export type DendroFile = {
    projectId: string
    fileId: string
    userId: string
    fileName: string
    size: number
    timestampCreated: number
    content: string
    metadata: any
    jobId?: string | null
}

export const isDendroFile = (x: any): x is DendroFile => {
    return validateObject(x, {
        projectId: isString,
        fileId: isString,
        userId: isString,
        fileName: isString,
        size: isNumber,
        timestampCreated: isNumber,
        content: isString,
        metadata: () => true,
        jobId: optional(isOneOf([isString, isNull]))
    })
}

export type ComputeResourceAwsBatchOpts = {
    jobQueue: string
    jobDefinition: string
}

export const isComputeResourceAwsBatchOpts = (x: any): x is ComputeResourceAwsBatchOpts => {
    return validateObject(x, {
        jobQueue: isString,
        jobDefinition: isString
    })
}


export type ComputeResourceSlurmOpts = {
    partition?: string
    time?: string
    cpusPerTask?: number
    otherOpts?: string
}

export const isComputeResourceSlurmOpts = (x: any): x is ComputeResourceSlurmOpts => {
    return validateObject(x, {
        partition: optional(isString),
        time: optional(isString),
        cpusPerTask: optional(isNumber),
        otherOpts: optional(isString)
    })
}

export type DendroComputeResourceApp = {
    name: string
    specUri?: string
    executablePath?: string // to be removed
    container?: string // to be removed
    awsBatch?: ComputeResourceAwsBatchOpts
    slurm?: ComputeResourceSlurmOpts
}

export const isDendroComputeResourceApp = (x: any): x is DendroComputeResourceApp => {
    return validateObject(x, {
        name: isString,
        specUri: optional(isString),
        executablePath: optional(isString), // to be removed
        container: optional(isString), // to be removed
        awsBatch: optional(isComputeResourceAwsBatchOpts),
        slurm: optional(isComputeResourceSlurmOpts)
    })
}

export type ComputeResourceSpecApp = {
    name: string
    help: string
    processors: ComputeResourceSpecProcessor[]
    appImage?: string
    appExecutable?: string
}

export const isComputeResourceSpecApp = (x: any): x is ComputeResourceSpecApp => {
    return validateObject(x, {
        name: isString,
        help: isString,
        processors: isArrayOf(isComputeResourceSpecProcessor),
        appImage: isString,
        appExecutable: isString
    })
}

export type ComputeResourceSpec = {
    apps: ComputeResourceSpecApp[]
}

export const isComputeResourceSpec = (x: any): x is ComputeResourceSpec => {
    return validateObject(x, {
        apps: isArrayOf(isComputeResourceSpecApp)
    })
}

export type DendroComputeResource = {
    computeResourceId: string
    ownerId: string
    name: string
    timestampCreated: number
    apps: DendroComputeResourceApp[]
    spec?: ComputeResourceSpec
}

export const isDendroComputeResource = (x: any): x is DendroComputeResource => {
    return validateObject(x, {
        computeResourceId: isString,
        ownerId: isString,
        name: isString,
        timestampCreated: isNumber,
        apps: isArrayOf(isDendroComputeResourceApp),
        spec: optional(isComputeResourceSpec)
    })
}

export type PubsubSubscription = {
    pubnubSubscribeKey: string
    pubnubChannel: string
    pubnubUser: string
}

export const isPubsubSubscription = (x: any): x is PubsubSubscription => {
    return validateObject(x, {
        pubnubSubscribeKey: isString,
        pubnubChannel: isString,
        pubnubUser: isString
    })
}

export type ProcessorGetJobResponseInput = {
    name: string
    url: string
}

export const isProcessorGetJobResponseInput = (x: any): x is ProcessorGetJobResponseInput => {
    return validateObject(x, {
        name: isString,
        url: isString
    })
}

export type ProcessorGetJobResponseOutput = {
    name: string
}

export const isProcessorGetJobResponseOutput = (x: any): x is ProcessorGetJobResponseOutput => {
    return validateObject(x, {
        name: isString
    })
}

export type ProcessorGetJobResponseParameter = {
    name: string
    value: any
}

export const isProcessorGetJobResponseParameter = (x: any): x is ProcessorGetJobResponseParameter => {
    return validateObject(x, {
        name: isString,
        value: isString
    })
}

export type ProcessorGetJobResponse = {
    jobId: string
    status: string
    processorName: string
    inputs: ProcessorGetJobResponseInput[]
    outputs: ProcessorGetJobResponseOutput[]
    parameters: ProcessorGetJobResponseParameter[]
}

export const isProcessorGetJobResponse = (x: any): x is ProcessorGetJobResponse => {
    return validateObject(x, {
        jobId: isString,
        status: isString,
        processorName: isString,
        inputs: isArrayOf(isProcessorGetJobResponseInput),
        outputs: isArrayOf(isProcessorGetJobResponseOutput),
        parameters: isArrayOf(isProcessorGetJobResponseParameter)
    })
}

File: python/dendro/common/dendro_types.py
from typing import Union, List, Any, Optional, Literal

from .. import BaseModel

class DendroProjectUser(BaseModel):
    userId: str
    role: str # 'admin' | 'editor' | 'viewer'

class DendroProject(BaseModel):
    projectId: str
    name: str
    description: str
    ownerId: str
    users: List[DendroProjectUser]
    publiclyReadable: bool
    tags: List[str]
    timestampCreated: float
    timestampModified: float
    computeResourceId: Union[str, None] = None # it seems this needs to go at the end, otherwise it will be required by pydantic - not sure why
    analysisSourceUrl: Union[str, None] = None

class DendroJobInputFile(BaseModel):
    name: str
    fileId: str
    fileName: str
    isFolder: Union[bool, None] = None

class DendroJobInputParameter(BaseModel):
    name: str
    value: Union[Any, None] = None
    secret: Union[bool, None] = None

class DendroJobOutputFile(BaseModel):
    name: str
    fileName: str
    fileId: Union[str, None] = None
    isFolder: Union[bool, None] = None

class ComputeResourceSpecProcessorParameter(BaseModel):
    name: str
    description: str
    type: str
    default: Union[Any, None] = None
    options: Union[List[str], List[int], None] = None
    secret: bool = False

class ComputeResourceSpecProcessorInput(BaseModel):
    name: str
    description: str
    list: bool = False

class ComputeResourceSpecProcessorInputFolder(BaseModel):
    name: str
    description: str
    list: bool = False

class ComputeResourceSpecProcessorOutput(BaseModel):
    name: str
    description: str

class ComputeResourceSpecProcessorOutputFolder(BaseModel):
    name: str
    description: str

class ComputeResourceSpecProcessorAttribute(BaseModel):
    name: str
    value: Any

class ComputeResourceSpecProcessorTag(BaseModel):
    tag: str

class ComputeResourceSpecProcessor(BaseModel):
    name: str
    description: str
    inputs: List[ComputeResourceSpecProcessorInput]
    inputFolders: Union[List[ComputeResourceSpecProcessorInputFolder], None] = None
    outputs: List[ComputeResourceSpecProcessorOutput]
    outputFolders: Union[List[ComputeResourceSpecProcessorOutputFolder], None] = None
    parameters: List[ComputeResourceSpecProcessorParameter]
    attributes: List[ComputeResourceSpecProcessorAttribute]
    tags: List[ComputeResourceSpecProcessorTag]

class DendroJobRequiredResources(BaseModel):
    numCpus: int
    numGpus: int
    memoryGb: float
    timeSec: float

class DendroJobUsedResources(BaseModel):
    numCpus: int
    numGpus: int
    memoryGb: float
    timeSec: float

class DendroJob(BaseModel):
    projectId: str
    jobId: str
    jobPrivateKey: str
    userId: str
    processorName: str
    inputFiles: List[DendroJobInputFile]
    inputFileIds: List[str]
    inputParameters: List[DendroJobInputParameter]
    outputFiles: List[DendroJobOutputFile]
    requiredResources: Union[DendroJobRequiredResources, None] = None
    usedResources: Union[DendroJobUsedResources, None] = None
    runMethod: Union[Literal['local', 'aws_batch', 'slurm'], None] = None
    timestampCreated: float
    computeResourceId: str
    status: str # 'pending' | 'queued' | 'starting' | 'running' | 'completed' | 'failed'
    batchId: Union[str, None] = None
    error: Union[str, None] = None
    processorVersion: Union[str, None] = None
    computeResourceNodeId: Union[str, None] = None # obsolete
    computeResourceNodeName: Union[str, None] = None # obsolete
    consoleOutputUrl: Union[str, None] = None
    resourceUtilizationLogUrl: Union[str, None] = None
    timestampQueued: Union[float, None] = None
    timestampStarting: Union[float, None] = None
    timestampStarted: Union[float, None] = None
    timestampFinished: Union[float, None] = None
    outputFileIds: Union[List[str], None] = None
    processorSpec: ComputeResourceSpecProcessor
    dandiApiKey: Union[str, None] = None
    deleted: Union[bool, None] = None
    pendingApproval: Union[bool, None] = None

class DendroFile(BaseModel):
    projectId: str
    fileId: str
    userId: str
    fileName: str
    size: int
    timestampCreated: float
    content: str # e.g., 'url:https://...'
    metadata: dict
    isFolder: Union[bool, None] = None
    jobId: Union[str, None] = None # the job that produced this file

# obsolete
class ComputeResourceAwsBatchOpts(BaseModel):
    jobQueue: Optional[str] = None # obsolete
    jobDefinition: Optional[str] = None # obsolete
    useAwsBatch: Optional[bool] = None # obsolete

# obsolete
class ComputeResourceSlurmOpts(BaseModel):
    partition: Union[str, None] = None
    time: Union[str, None] = None
    cpusPerTask: Union[int, None] = None
    otherOpts: Union[str, None] = None

class DendroComputeResourceApp(BaseModel):
    name: str
    specUri: str
    executablePath: Union[str, None] = None # to be removed (once database has been cleared)
    container: Union[str, None] = None # to be removed (once database has been cleared)
    awsBatch: Union[ComputeResourceAwsBatchOpts, None] = None # obsolete
    slurm: Union[ComputeResourceSlurmOpts, None] = None # obsolete

class ComputeResourceSpecApp(BaseModel):
    name: str
    description: str
    processors: List[ComputeResourceSpecProcessor]
    appImage: Union[str, None] = None
    appExecutable: Union[str, None] = None

class ComputeResourceSpec(BaseModel):
    apps: List[ComputeResourceSpecApp]
    defaultJobRunMethod: Union[Literal['local', 'aws_batch', 'slurm'], None] = None
    availableJobRunMethods: Union[List[Literal['local', 'aws_batch', 'slurm']], None] = None

class DendroComputeResource(BaseModel):
    computeResourceId: str
    ownerId: str
    name: str
    timestampCreated: float
    apps: List[DendroComputeResourceApp]
    spec: Union[ComputeResourceSpec, None] = None

class PubsubSubscription(BaseModel):
    pubnubSubscribeKey: str
    pubnubChannel: str
    pubnubUser: str

class ProcessorGetJobResponseInput(BaseModel):
    name: str
    url: str

class ProcessorGetJobResponseInputFolderFile(BaseModel):
    name: str
    url: str
    size: Union[int, None] = None

class ProcessorGetJobResponseInputFolder(BaseModel):
    name: str
    files: List[ProcessorGetJobResponseInputFolderFile]

class ProcessorGetJobResponseOutput(BaseModel):
    name: str

class ProcessorGetJobResponseOutputFolder(BaseModel):
    name: str

class ProcessorGetJobResponseParameter(BaseModel):
    name: str
    value: Any

class ProcessorGetJobResponse(BaseModel):
    jobId: str
    status: str
    processorName: str
    inputs: List[ProcessorGetJobResponseInput]
    inputFolders: Union[List[ProcessorGetJobResponseInputFolder], None] = None
    outputs: List[ProcessorGetJobResponseOutput]
    outputFolders: Union[List[ProcessorGetJobResponseOutputFolder], None] = None
    parameters: List[ProcessorGetJobResponseParameter]

class DendroUser(BaseModel):
    userId: str
    dendroApiKey: Union[str, None] = None

class CreateJobRequestInputFile(BaseModel):
    name: str
    fileName: str
    isFolder: Union[bool, None] = None

class CreateJobRequestOutputFile(BaseModel):
    name: str
    fileName: str
    isFolder: Union[bool, None] = None

class CreateJobRequestInputParameter(BaseModel):
    name: str
    value: Union[Any, None]

class CreateJobRequest(BaseModel):
    projectId: str
    processorName: str
    inputFiles: List[CreateJobRequestInputFile]
    outputFiles: List[CreateJobRequestOutputFile]
    inputParameters: List[CreateJobRequestInputParameter]
    processorSpec: ComputeResourceSpecProcessor
    batchId: Union[str, None] = None
    dandiApiKey: Union[str, None] = None
    requiredResources: DendroJobRequiredResources
    runMethod: Literal['local', 'aws_batch', 'slurm']

class CreateJobResponse(BaseModel):
    jobId: str
    success: bool

class ComputeResourceUserUsage(BaseModel):
    computeResourceId: str
    userId: str
    jobsIncludingDeleted: List[DendroJob]
